### PR TITLE
[frontend] Show artifact preview in UI (#2172)

### PR DESCRIPTION
**Pull request by eterna2**
Part of #2172. This PR removes the MinioArtifactLink component and replaces it with a MinioArtifactPreview. MinioArtifactPreview takes in an Argo S3Artifact object and attempts to load and render a preview to the artifact content. If the attempts fails, no preview will be shown, but the error will be logged to console.

A screenshot of the UI is shown below:
artifactpreview

artifactpreview2

cc: @Bobgy

### CODE REVIEW

**Comment by Bobgy** [comment]
@eterna2 This looks great!

Can you add a screenshot when there are more than one artifacts?

File: frontend/src/lib/Utils.tsx [hunk 1]
```diff
@@ -292,8 +292,23 @@ export function generateGcsConsoleUri(gcsUri: string): string | undefined {
...
+ * @param peek number of characters or bytes to return. If not provided, the entire content of the artifact will be returned.
+ */
+export function generateArtifactUrl(
+  source: string,
+  bucket: string,
+  key: string,
+  peek?: number,
+): string {
+  return encodeURI(
+    `artifacts/get?source=${source}&bucket=${bucket}&key=${key}${peek ? `&peek=${peek}` : ''}`,
```
[the remark this answers is not in the document]
Comment (eterna2) [reply frontend/src/lib/Utils.tsx:310]: i intentionally did not use buildQuery because of the encodeUriComponent. Because I think the keys and bucket etc from the S3 artifact is already encoded. 

In my prev PR, i created the consistent encode and decode because of this issue - double encoded in some cases. 

But let me make the change and try again.

[the remark this answers is not in the document]
Comment (Bobgy) [reply frontend/src/lib/Utils.tsx:310]: I see, if there's something blocking that, feel free to leave it for future improvements.

**Comment by Bobgy** [comment]
@eterna2 note you can ignore travis tests when developing, it's kind of flaky recently. And even if you retry them, they always fail.
Pushing extra empty commits can retry them.

File: frontend/src/lib/Utils.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (eterna2) [reply frontend/src/lib/Utils.tsx]: ok. I did a test run, so far all the workflow artifacts and ml-metadata artifacts look fine too.

**Comment by eterna2** [comment]
ok i fixed everything.

File: frontend/src/components/DetailsTable.tsx [hunk 2]
```diff
@@ -97,18 +103,16 @@ const DetailsTable = (props: DetailsTableProps) => {
               // do nothing
             }
           }
-          // If the value isn't a JSON object, just display it as is
+          // If value is a S3Artifact render a preview, otherwise just display it as is
```
Comment (Bobgy) [inline frontend/src/components/DetailsTable.tsx:106]: nit: update the comment?

File: frontend/src/components/MinioArtifactPreview.test.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ */
+import React from 'react';
+import { render, act } from '@testing-library/react';
+import MinioArtifactPreview from './MinioArtifactPreview';
+import { Apis } from '../lib/Apis';
+
+jest.mock('../lib/Apis');
+
+describe('MinioArtifactPreview', () => {
+  const readFile = Apis.readFile as jest.Mock;
```
Comment (Bobgy) [inline frontend/src/components/MinioArtifactPreview.test.tsx:24]: a type safe way to get the spy is
```
const readFileSpy = jest.spyOn(Apis, 'readFile');
```
You can remove the `jest.mock` usage.

File: frontend/src/components/MinioArtifactPreview.test.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      bucket: 'foo',
+      endpoint: 's3.amazonaws.com',
+      key: 'bar',
+      secretKeySecret: { key: 'secretkey', optional: false, name: 'minio' },
+    };
+
+    const container = document.body.appendChild(document.createElement('div'));
+    await act(async () => {
+      render(<MinioArtifactPreview artifact={s3Artifact} />, { container });
+    });
```
Comment (Bobgy) [inline frontend/src/components/MinioArtifactPreview.test.tsx:85]: This is kind of strange. Why did you need to create the div manually and wrap `render` in an `act`?

Comment (Bobgy) [reply frontend/src/components/MinioArtifactPreview.test.tsx:85]: Did you want to try something like https://github.com/kubeflow/pipelines/blob/f882f361cc705580261128e33523fe80d781b30f/frontend/src/components/PodYaml.test.tsx#L37? to wait for promise to resolve before finally rendering the container?

File: frontend/src/lib/Apis.ts [hunk 5]
```diff
@@ -205,9 +206,28 @@ export class Apis {
...
-    return this._fetch(`artifacts/get${buildQuery({ source, bucket, key, namespace })}`);
+  public static readFile(path: StoragePath, namespace?: string, peek?: number): Promise<string> {
+    return this._fetch(this.buildReadFileUrl(path, namespace, peek));
+  }
+
+  /**
+   * Builds an url for the readFile API to retrieve a workflow artifact.
+   * @param props object describing the artifact (e.g. source, bucket, and key)
+   */
+  public static buildReadFileUrl(path: StoragePath, namespace?: string, peek?: number) {
```
Comment (Bobgy) [inline frontend/src/lib/Apis.ts:217]: Awesome! This looks great

File: frontend/src/components/DetailsTable.tsx [hunk 6]
```diff
@@ -48,22 +47,29 @@ export const css = stylesheet({
...
+  value?: string | T;
+  [key: string]: any;
+}
+
+interface DetailsTableProps<T> {
+  fields: Array<KeyValue<string | T>>;
   title?: string;
-  valueComponent?: React.FC<{ value: S3Artifact }>;
+  valueComponent?: React.FC<ValueComponentProps<T>>;
+  [key: string]: any;
```
Comment (Bobgy) [inline frontend/src/components/DetailsTable.tsx:59]: Thanks! this looks like a good approach.
nit: probably slightly change it to
```
valueComponentProps: { [key: string]: any };
```
to make it more focused.

Otherwise, DetailsTable will accept any key as a prop, which is not good for type checking.

**Comment by eterna2** [comment]
ok fixed everything.

- used `valueComponentProps` to pass the extra props to the valueComponent
- fixed the unit test with `act`
- fixed the comments
- fixed the handling of invalid s3artifact (now it return null also)
- add unit test if value is string (or other primitive) -> render as string (i.e. MinioArtifactPreview will render valid s3artifact values properly, invalid objects as null, undefined/null values as null, otherwise render as a string)

**Comment by eterna2** [comment]
hmmm... got a unit test error
```
FAIL src/pages/RunDetails.test.tsx (41.301s)
  ● Console
    console.warn node_modules/react-dom/cjs/react-dom.development.js:12357
      Warning: componentWillMount has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.
```

**Comment by eterna2** [comment]
the strangest part is that only `RunDetails switches to inputs/outputs tab in side pane 1` shld be showing the `DetailsTable`. 

not `RunDetails opens side panel when graph node is clicked 1`

**Comment by eterna2** [comment]
ok i found the issue. ur recent update to the visualization tab changed the snapshot. So the unit test is testing against that. 

I have rebase to the latest master and updated the snapshot. Should work now.

File: frontend/src/components/MinioArtifactPreview.tsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  namespace,
+  peek = 255,
+}) => {
+  const [content, setContent] = React.useState<string | undefined>(undefined);
+  const storagePath = getStoragePath(value);
+
+  React.useEffect(() => {
+    let cancelled = false;
+    if (storagePath) {
+      // TODO how to handle binary data
```
Comment (Ark-kun) [inline frontend/src/components/MinioArtifactPreview.tsx:61]: Later, we can abort preview if unprintable bytes are found.

File: frontend/src/components/MinioArtifactPreview.tsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      Apis.readFile(storagePath, namespace, peek + 1).then(
+        data => !cancelled && setContent(data.length > peek ? `${data.slice(0, peek)} ...` : data),
```
Comment (Ark-kun) [inline frontend/src/components/MinioArtifactPreview.tsx:63]: Later we can have more advanced logic here. For example <=10 full lines or <= 10KB whichever comes first.

P.S. The ellipsis is a nice touch (maybe later we can put it on its own line)

**Comment by Ark-kun** [comment]
Thank you a lot for this PR. This really improves the experience!

File: frontend/src/lib/Apis.ts [hunk 9]
```diff
@@ -205,9 +205,28 @@ export class Apis {
...
+  }
+
+  /**
+   * Builds an url to visually represents a workflow artifact location.
+   * @param param.source source of the artifact (e.g. minio, gcs, s3, http, or https)
+   * @param param.bucket name of the bucket with the artifact (or host for http/https)
+   * @param param.key key (i.e. path) of the artifact in the bucket
+   */
+  public static buildArtifactUrl({ source, bucket, key }: StoragePath) {
+    // TODO see https://github.com/kubeflow/pipelines/pull/3725
```
Comment (Ark-kun) [inline frontend/src/lib/Apis.ts:228]: For now, I've decided to put the Argo artifact info to the MLMD artifact custom property. So there will be no new URIs for now.

File: frontend/src/components/DetailsTable.test.tsx [hunk 10]
```diff
@@ -55,60 +101,502 @@ describe('DetailsTable', () => {
...
+        <div>
+          <div
+            class="row"
+          >
+            <span
+              class="key"
+            >
+              key
+            </span>
+            <div
```
Comment (Bobgy) [inline frontend/src/components/DetailsTable.test.tsx:173]: nit: snapshot for ace editor is long and it didn't even include the actual content.
Recommend mocking it out like https://github.com/Bobgy/pipelines/blob/68c55e035640be29f89b8f4896d045acf6595c91/frontend/src/components/PodYaml.test.tsx#L7

File: frontend/src/components/MinioArtifactPreview.test.tsx [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    expect(container).toMatchInlineSnapshot(`
+      <div>
+        <div
+          class="root"
+        >
+          <a
+            class="link"
+            rel="noopener"
+            target="_blank"
+          />
```
Comment (Bobgy) [inline frontend/src/components/MinioArtifactPreview.test.tsx:219]: All the <a/>s should have href and content, is this a bug?

File: frontend/src/components/MinioArtifactPreview.tsx (hunk 8 above)
Comment (Bobgy) [reply frontend/src/components/MinioArtifactPreview.tsx:63]: +1 to putting `...` on its own line, why not do it now?

File: frontend/src/components/MinioArtifactPreview.test.tsx (hunk 11 above)
Comment (eterna2) [reply frontend/src/components/MinioArtifactPreview.test.tsx:219]: yes. I missed this too. Let me check

Comment (eterna2) [reply frontend/src/components/MinioArtifactPreview.test.tsx:219]: found the bug. cuz I mocked the entire `APIs` module so the functions did not return anything. Removed the unneccessary mocks. Works now.

File: frontend/src/components/DetailsTable.test.tsx (hunk 10 above)
Comment (eterna2) [reply frontend/src/components/DetailsTable.test.tsx]: ok done

**Comment by eterna2** [comment]
ok made the fixes.

Also changed the props from `peek` to `maxbytes` and `maxlines`. 
`maxlines` is optional.

File: frontend/src/components/MinioArtifactPreview.test.tsx [hunk 12]
```diff
@@ -1,3 +1,8 @@
+import MinioArtifactPreview from './MinioArtifactPreview';
```
Comment (Bobgy) [inline frontend/src/components/MinioArtifactPreview.test.tsx:1]: nit: Can you move imports under license declaration?

Comment (eterna2) [reply frontend/src/components/MinioArtifactPreview.test.tsx:1]: ops. ide auto import hook not smart enuff to recognize license

Comment (eterna2) [reply frontend/src/components/MinioArtifactPreview.test.tsx]: done

**Comment by Bobgy** [comment]
Awesome! Really useful feature!

/lgtm

**Comment by Bobgy** [comment]
/hold
unfortunately, travis tests failed
Can you submit an additional commit with any change to retry it?

**Comment by Bobgy** [comment]
/cc @Ark-kun 
I've seen travis tests being flaky and unretriable quite often recently. Shall we consider changing it to non required and approver/lgtmer should check it passes?

**Comment by Bobgy** [comment]
Hmmm, it failed again. I'll help you merge this PR directly.

**Comment by eterna2** [comment]
I looked at the logs. The Travis build was terminated because it timeout after 10 minutes.

I think u can go increase the timeout

**Comment by eterna2** [comment]
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated
```

**Comment by Bobgy** [comment]
> I looked at the logs. The Travis build was terminated because it timeout after 10 minutes.
> 
> I think u can go increase the timeout

Thanks, that's interesting.
Can you give a screenshot where you saw that?

**Comment by Bobgy** [comment]
I was looking at the error in https://travis-ci.com/github/kubeflow/pipelines/jobs/331832190#L3086.
It shows up randomly for all three tests on different python versions.

**Comment by eterna2** [comment]
> > I looked at the logs. The Travis build was terminated because it timeout after 10 minutes.
> > I think u can go increase the timeout
> 
> Thanks, that's interesting.
> Can you give a screenshot where you saw that?

here
https://travis-ci.com/github/kubeflow/pipelines/jobs/331472823#L3205

**Comment by Bobgy** [comment]
@eterna2 Thanks!
If you check PR travis test history here: https://travis-ci.com/github/kubeflow/pipelines/pull_requests. All other failures seem to be the same as the one I posted above. There was only one time it showed time out. We'd focus on solving the problem in https://github.com/kubeflow/pipelines/pull/3707#issuecomment-626652201 with higher priority.
/cc @Ark-kun

**Comment by Ark-kun** [comment]
> I've seen travis tests being flaky and unretriable quite often recently.

The Travis errors are more retriable now.
There were several issues recently:
* Travis' "No output has been received in the last 10m0s". Happens rarely, but is retriable
* Coveralls error 422. This was an unretriable error. I've removed coveralls call.
* TFX installation failures. Unretriable. Fixed on the TFX side.
* TFX installation flakyness. Somewhat retriable. See my issue https://github.com/pypa/pip/issues/8218

So, only retriable flakiness remains. Sorry for the inconvenience.

**Comment by Bobgy** [comment]
> > I've seen travis tests being flaky and unretriable quite often recently.
> 
> The Travis errors are more retriable now.
> There were several issues recently:
> 
> * Travis' "No output has been received in the last 10m0s". Happens rarely, but is retriable
> * Coveralls error 422. This was an unretriable error. I've removed coveralls call.
> * TFX installation failures. Unretriable. Fixed on the TFX side.
> * TFX installation flakyness. Somewhat retriable. See my issue [pypa/pip#8218](https://github.com/pypa/pip/issues/8218)
> 
> So, only retriable flakiness remains. Sorry for the inconvenience.

@Ark-kun thanks for fixing them, that's great news for me.

File: frontend/src/components/MinioArtifactPreview.tsx [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    // if value is undefined, null, or an invalid s3artifact object, don't render
+    if (value === null || value === undefined || typeof value === 'object') return null;
+    // otherwise render value as string (with default string method)
+    return <React.Fragment>{`${value}`}</React.Fragment>;
+  }
+
+  // TODO need to come to an agreement how to encode artifact info inside a url
+  // namespace is currently not supported
+  const linkText = Apis.buildArtifactUrl(storagePath);
+  const artifactUrl = Apis.buildReadFileUrl(storagePath, namespace, maxbytes);
```
Comment (Bobgy) [inline frontend/src/components/MinioArtifactPreview.tsx:108]: @eterna2 I overlooked a bug. Here, maxbytes shouldn't be set, because when we click the artifact link, we should expect user wants to see full content, is that right?

Comment (eterna2) [reply frontend/src/components/MinioArtifactPreview.tsx:108]: my bad. I missed this too.
